vmupdate/qube_connection.py

```diff
@@ -0,0 +1,233 @@
+        output += output_
+
+        return exit_code, output
+
+    def _copy_file_from_dom0(self, src, dest):
+        read_src = ["cat", src]
+        qvm_run = ["qvm-run", "--user=root", "--pass-io", self.qube.name]
+        write_dest = ["cat", ">", dest]
+        command = [*qvm_run, " ".join(write_dest)]
+        self.logger.debug(
+            "run command: %s | %s", " ".join(read_src), " ".join(command))
+        try:
+            p1 = Popen(read_src, stdout=subprocess.PIPE)
+            p2 = Popen(command, stdin=p1.stdout)
+            p1.stdout.close()  # Allow p1 to receive a SIGPIPE if p2 exits.
+            p2.communicate()
+            p1.wait()  # wait for returncode
+            ret_code = max(p1.returncode, p2.returncode)
+            if ret_code:
+                raise OSError(f"Command returns code: {ret_code}")
+            output = ""
+        except OSError as exc:
+            ret_code = 1
+            output = str(exc)
+
+        return ret_code, output
+
+    def run_entrypoint(self, entrypoint_path, agent_args):
+        """
+        Run a script in the qube.
+
+        :param entrypoint_path: str: path to the entrypoint.py in the qube
+        :param agent_args: args for agent entrypoint
+        :return: Tuple[int, str]: return code and output of the script
+        """
+        # make sure entrypoint is executable
+        command = ['chmod', 'u+x', entrypoint_path]
+        exit_code, output = self._run_shell_command_in_qube(self.qube, command)
+
+        # run entrypoint
+        command = [entrypoint_path, *AgentArgs.to_cli_args(agent_args)]
+        exit_code_, output_ = self._run_shell_command_in_qube(
+            self.qube, command, show=self.show_progress
+        )
+        exit_code = max(exit_code, exit_code_)
+        output += output_
+
+        return exit_code, output
+
+    def read_logs(self):
+        """
+        Read vm logs file.
+        """
+        command = ['cat',
+                   str(join(package_manager.LOGPATH, package_manager.LOG_FILE))]
+        exit_code, output = self._run_shell_command_in_qube(self.qube, command)
+        return exit_code, output
+
+    def _run_shell_command_in_qube(self, target, command, show=False):
+        self.logger.debug("run command in %s: %s",
+                          target.name, " ".join(command))
+        if not show:
+            try:
+                untrusted_stdout_and_stderr = target.run_with_args(*command,
+                                                                   user='root')
+                ret_code = 0
+            except CalledProcessError as e:
+                self.logger.error(str(e))
+                ret_code = e.returncode
+                untrusted_stdout_and_stderr = (b"", b"")
+        else:
+            p = target.run_service(
+                'qubes.VMExec+' + qubesadmin.utils.encode_for_vmexec(command),
+                user='root')
+            p.stdin.write((" ".join(command) + "\n").encode())
```

Comment:
writing command here is not necessary anymore - you already sent it the line above